### ISSUE TITLE
Template argument special variables within typemaps

### ISSUE THREAD

**Issue by reductor** [swig/swig#1149]
It would be useful to have the ability to lookup template argument types within type this would make it easier to write typemaps for templates, and probably remove alot of the boilerplate code currently required.

An example of this would be
`%typemap(cstype) std::vector "System.Collections.Generic.List<$typemap(cstype,$1_tparam1)>"`

This would be able to translate
`std::vector<int>` into `System.Collections.Generic.List<int>`

Doing this I believe will remove the need for alot of the macros/defines which exist for handling C++ templates as the typemap lookup can just be done on the actual template type.

I already have a patch that provides this functionality, is there any issues that someone can see with this? Or should I just put up a pull request and it can be the discussion?

**Comment by wsfulton** [comment]
As an aside, you might be interested in #887 which has been merged in for the next release and may make the need for your custom typemap redundant as you can construct a std::vector<int> from any C# container.

**Comment by wsfulton** [comment]
Regarding the `$1_tparam1` idea, I am not in favour of this additional feature as there is a generic solution to access access template variables via %extend. Let's first consider a slightly different case to `std::vector<int>` and instead use `std::vector<X>` where X is a class:
```swig
%inline %{
struct X {};
%}

%include <std_vector.i>

%extend std::vector {
  %typemap(cstype) std::vector<T> "System.Collections.Generic.List<$typemap(cstype,T)>"
}

%template(VectorX) std::vector<X>;

%inline %{
void usevector(std::vector<X> vi) {}
%}
```
The typemap is found and used as intended in the generated wrappers:
```c#
public class example {
  public static void usevector(System.Collections.Generic.List<X> vi) {
...
```
This doesn't work straight off if you change X to int. This might at first be confusing but is correct as the std::vector definition in std_vector.i has a primary template which is used for X. There is a template specialization for `std::vector<int>` so the primary template is not used. The %extend above is for the primary template and so the typemap is not used. In a specialized template there is no `T`, just `int`, so the template parameter needs to be specified as int, so unfortunately in this case there is no shortcut and there is little point using %extend, so just use:
```swig
  %typemap(cstype) std::vector<int> "System.Collections.Generic.List<$typemap(cstype,int)>"
```
The C# std::vector class is one of the very few classes that has template specializations provided by SWIG and so you just need to provide the typemap for each of the specializations. As it is such an unusual case, it doesn't warrant complicating SWIG with additional special variables that only work inside of typemaps and not anywhere else. What we should do is document this technique more clearly.

**Comment by ReDucTor** [comment]
Thanks for the response, I'm still confused with how this can be achieved currently without using some sort of %define which must be applied to every template argument combination, another example usage is attempting to achieve something like the proposed std::expected<T>

https://gist.github.com/ReDucTor/84835c86737221de811b3cc9c2ccf885

How would the same thing be achieved using the generic %extend ?

**Comment by ReDucTor** [comment]
After doing some more fiddling I managed to get %extend to work but didn't appear to work for all typemaps

https://gist.github.com/ReDucTor/3fc34fa8a0528aa9911ce6beee40c3f4

**Comment by wsfulton** [comment]
Which parts are not working? Are you referring to the `imtype` typemap? If so, you are missing the out attribute in the typemap (see csharp.swg) and so you need to specify it:
```
  %typemap(imtype, out="global::System.IntPtr") expected<T> "$typemap(imtype,T)"
```
I can't see a way to set the out attribute by obtaining the out attribute from another typemap using `$typemap`. This is a shortcoming that could be fixed if this is what you were trying to do? We'd need to think of a suitable syntax for making this work though.

**Comment by ReDucTor** [comment]
It appears that it was working for the other typemaps, I was just not reading the output correct.

The other disadvantage of this approach is that I can't use %ignore for expected and still have it process the typemaps, is there some other way of having it disable the generation of the code for expected?

`$null` is the other thing which I also need to find another way of handling from the typemap, possibly it might be useful to be able to have something like

`%typemap(imtype, out="$typemap(imtype:out,T") expected<T> "$typemap(imtype,T)"`

It would be good to avoid having to use `%template` for each of these and have them implicitly generated any time they would be needed.

**Comment by ojwb** [comment]
@ReDucTor I think #2220 implements what you suggest in your most recent comment above.

**Comment by ojwb** [comment]
#2220 is now merged and was in SWIG 4.1.0 so this should now work:

```
%typemap(imtype, out="$typemap(imtype:out,T") expected<T> "$typemap(imtype,T)"
```

That seems to be the only outstanding point, so closing.  If I missed something, please say what and we can reopen.
